### PR TITLE
[API] Registering shop users

### PR DESCRIPTION
**Pull request by pamil**
We need to customise the test API client to not force `admin` prefix in further PRs.

### CODE REVIEW

File: src/Sylius/Bundle/ApiBundle/spec/CommandHandler/RegisterShopUserHandlerSpec.php [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+        $customer->setEmail('WILL.SMITH@example.com')->shouldBeCalled();
+        $customer->setPhoneNumber('+13104322400')->shouldBeCalled();
+        $customer->setUser($shopUser)->shouldBeCalled();
+
+        $shopUserManager->persist($shopUser)->shouldBeCalled();
+
+        $this->__invoke(new RegisterShopUser('Will', 'Smith', 'WILL.SMITH@example.com', 'iamrobot', '+13104322400'));
+    }
+
+    function it_creates_a_user_and_existing_customer_with_given_data(
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/spec/CommandHandler/RegisterShopUserHandlerSpec.php:77]:
```suggestion
    function it_creates_a_user_only_if_customer_already_exists(
```

File: src/Sylius/Bundle/ApiBundle/spec/CommandHandler/RegisterShopUserHandlerSpec.php [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+        $customer->setFirstName('Will')->shouldBeCalled();
+        $customer->setLastName('Smith')->shouldBeCalled();
+        $customer->setEmail('WILL.SMITH@example.com')->shouldBeCalled();
+        $customer->setPhoneNumber('+13104322400')->shouldBeCalled();
+        $customer->setUser($shopUser)->shouldBeCalled();
+
+        $shopUserManager->persist($shopUser)->shouldBeCalled();
+
+        $this->__invoke(new RegisterShopUser('Will', 'Smith', 'WILL.SMITH@example.com', 'iamrobot', '+13104322400'));
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/spec/CommandHandler/RegisterShopUserHandlerSpec.php:74]: Just as a proposal for different style (I prefer this one in invokables), but feel free to reject it
```suggestion
        $this(new RegisterShopUser('Will', 'Smith', 'WILL.SMITH@example.com', 'iamrobot', '+13104322400'));
```

File: src/Sylius/Bundle/ApiBundle/CommandHandler/RegisterShopUserHandler.php [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    public function __invoke(RegisterShopUser $command): void
+    {
+        $emailCanonical = $this->canonicalizer->canonicalize($command->email);
+
+        /** @var ShopUserInterface $user */
+        $user = $this->shopUserFactory->createNew();
+        $user->setPlainPassword($command->password);
+
+        /** @var CustomerInterface|null $customer */
+        $customer = $this->customerRepository->findOneBy(['emailCanonical' => $emailCanonical]);
```
Comment (pamil) [inline src/Sylius/Bundle/ApiBundle/CommandHandler/RegisterShopUserHandler.php:65]: If the customer already has a user assigned, we need to throw an exception.

File: config/packages/messenger.yaml [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+            # async: '%env(MESSENGER_TRANSPORT_DSN)%'
+            # failed: 'doctrine://default?queue_name=failed'
+            # sync: 'sync://'
+
+        routing:
+            # Route your messages to the transports
+            # 'App\Message\YourMessage': async
+
+        buses:
+            messenger.bus.default:
```
[the remark this answers is not in the document]
Comment (pamil) [reply config/packages/messenger.yaml:17]: I'm not sure if I should tackle this issue in this PR - what about creating an issue about that to keep track of it, not to grow this PR even bigger.

File: etc/travis/suites/common/before_install.sh [hunk 5]
```diff
@@ -5,6 +5,7 @@ source "$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)/../../../bash/common.lib.s
 print_header "Customizing the environment" "Sylius"
 run_command "git fetch origin $TRAVIS_BRANCH:refs/remotes/origin/$TRAVIS_BRANCH" || exit $? # Make origin/master available for is_suitable steps
 run_command "phpenv config-rm xdebug.ini" # Disable XDebug
-run_command "echo \"memory_limit=4096M\" >> ~/.phpenv/versions/$(phpenv version-name)/etc/conf.d/travis.ini" || exit $? # Increase memory limit to 4GB
+run_command "echo \"memory_limit=6144M\" >> ~/.phpenv/versions/$(phpenv version-name)/etc/conf.d/travis.ini" || exit $? # Increase memory limit to 4GB
```
Comment (lchrusciel) [inline etc/travis/suites/common/before_install.sh:8]:
```suggestion
run_command "echo \"memory_limit=6144M\" >> ~/.phpenv/versions/$(phpenv version-name)/etc/conf.d/travis.ini" || exit $? # Increase memory limit to 6GB
```

File: bin/install-packages [hunk 6]
```diff
@@ -13,4 +13,4 @@ reset_cache() {
 
 install_package="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)/install-package"
 
-locate_packages | parallel -j "$(get_number_of_jobs_for_parallel)" --gnu "${install_package} {} $* $(reset_cache)"
+locate_packages | parallel -j1 --gnu "${install_package} {} $* $(reset_cache)"
```
Comment (lchrusciel) [inline bin/install-packages:16]: Why do we need this change?

File: src/Sylius/Bundle/ApiBundle/CommandHandler/RegisterShopUserHandler.php [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+use Doctrine\Persistence\ObjectManager;
+use Sylius\Bundle\ApiBundle\Command\RegisterShopUser;
+use Sylius\Component\Core\Model\CustomerInterface;
+use Sylius\Component\Core\Model\ShopUserInterface;
+use Sylius\Component\Core\Repository\CustomerRepositoryInterface;
+use Sylius\Component\Resource\Factory\FactoryInterface;
+use Sylius\Component\User\Canonicalizer\CanonicalizerInterface;
+use Symfony\Component\Messenger\Handler\MessageHandlerInterface;
+
+class RegisterShopUserHandler implements MessageHandlerInterface
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/CommandHandler/RegisterShopUserHandler.php:25]: I thought about opening  `RegisterShopUser` only. Imho, this class could be finalized

**Comment by lchrusciel** [comment]
Thank you, Kamil! :1st_place_medal:

**Pull request merged**